### PR TITLE
Smithy resources for dynamic module

### LINKED ISSUE #687: Model resources in dynamic module + handle them the same way codegen does
See https://github.com/disneystreaming/smithy4s/pull/686

When a service reference a resource, the operations of the resource (and all its sub-resources) should be added to the dynamic service. 

This implies that resources should be fully modelled in our dynamic model.

### PR DESCRIPTION
**Pull request by msosnicki**
Closes #687

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -204,24 +204,25 @@ lazy val core = projectMatrix
...
     Test / smithySpecs := Seq(
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "metadata.smithy",
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "recursive.smithy",
+      (ThisBuild / baseDirectory).value / "sampleSpecs" / "adtMember.smithy",
       (ThisBuild / baseDirectory).value / "sampleSpecs" / "bodies.smithy",
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "misc.smithy",
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "product.smithy",
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "weather.smithy",
+      (ThisBuild / baseDirectory).value / "sampleSpecs" / "defaults.smithy",
       (ThisBuild / baseDirectory).value / "sampleSpecs" / "discriminated.smithy",
```
Comment (msosnicki) [inline build.sbt:210]: Just sorted and added `resources.smithy`

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/OperationSpec.scala [hunk 2]
```diff
@@ -18,31 +18,12 @@ package smithy4s
...
 
-class OperationSpec() extends munit.FunSuite {
-
-  // This is not ideal, but it does the job.
-  val cwd = System.getProperty("user.dir");
-  val pizzaSpec = Paths.get(cwd + "/sampleSpecs/pizza.smithy").toAbsolutePath()
+class OperationSpec() extends DummyIO.Suite {
 
   test("Decode operation") {
-    IO(
```
Comment (msosnicki) [inline modules/dynamic/test/src-jvm/smithy4s/dynamic/OperationSpec.scala:35]: `check` no longer required (due to `DummyIO.Suite`). Here it was not called at all, but test was silently failing (missing `alloy` dependency).

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/package.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import smithy4s.dynamic.DummyIO._
+import java.nio.file.Paths
+import software.amazon.smithy.model.{Model => SModel}
+import software.amazon.smithy.model.shapes.ModelSerializer
+import cats.syntax.all._
+
+package object dynamic {
+  // This is not ideal, but it does the job.
+  private val cwd = System.getProperty("user.dir");
```
Comment (Baccata) [inline modules/dynamic/test/src-jvm/smithy4s/dynamic/package.scala]: make this lazy : running effects during instantiation of static objects (including package objects) is a big no-no

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/ResourceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  test("Compile resource operations as endpoints") {
+    loadDynamicModel("resources.smithy").map { model =>
+      val compiled = DynamicSchemaIndex.load(model)
+
+      val endpoints = compiled.allServices.head.service.endpoints
+      expect.same(
+        endpoints.map(_.id).toSet,
+        List("ListPublishers", "GetBook", "BuyBook")
```
Comment (msosnicki) [inline modules/dynamic/test/src-jvm/smithy4s/dynamic/ResourceSpec.scala]: What would be better ways to test it? Throw it at some algebra, like in `DynamicJsonServerSpec`?

File: build.sbt [hunk 5]
```diff
@@ -204,24 +204,25 @@ lazy val core = projectMatrix
       "smithy4s.example.collision"
     ),
     Test / smithySpecs := Seq(
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "metadata.smithy",
-      (ThisBuild / baseDirectory).value / "sampleSpecs" / "recursive.smithy",
+      (ThisBuild / baseDirectory).value / "sampleSpecs" / "adtMember.smithy",
```
Comment (Baccata) [inline build.sbt:207]: I appreciate the sorting, thanks

File: modules/dynamic/smithy/dynamic.smithy [hunk 6]
```diff
@@ -184,11 +184,20 @@ structure ServiceShape {
   version: String,
   errors: MemberList,
   operations: MemberList,
+  resources: MemberList,
   traits: TraitMap
 }
 
-/// TODO
 structure ResourceShape {
+  /// ignored: identifiers, properties, collectionOperations
```
Comment (msosnicki) [inline modules/dynamic/smithy/dynamic.smithy:192]: Couldn't get `collectionOperations` to work. Same applied for `list` member shape. Looks like they have some specific requirements, as I get following error: 
```
smithy4s.example#ListPublishers: This operation is bound as a collection operation on the `smithy4s.example#Publisher` resource, but it improperly binds all of the identifiers of the resource to members of the operation input. 
```
Will try to solve that and push a fix for it

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/package.scala (hunk 3 above)
Comment (msosnicki) [reply modules/dynamic/test/src-jvm/smithy4s/dynamic/package.scala]: Good catch, will do!

File: modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala [hunk 7]
```diff
@@ -25,15 +26,20 @@ object DummyIO {
...
-    def check(): Unit = io match {
-      case Left(e)  => throw e
-      case Right(_) => ()
-    }
+
+  trait Suite extends munit.FunSuite {
+    override def munitValueTransforms: List[ValueTransform] =
+      ioValueTransform :: super.munitValueTransforms
+
+    private val ioValueTransform: ValueTransform =
```
Comment (Baccata) [inline modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala:34]: 👍

File: modules/example/src/smithy4s/example/GetObjectOutput.scala [hunk 8]
```diff
@@ -14,7 +14,7 @@ object GetObjectOutput extends ShapeTag.Companion[GetObjectOutput] {
   val hints: Hints = Hints.empty
 
   implicit val schema: Schema[GetObjectOutput] = struct(
-    ObjectSize.schema.required[GetObjectOutput]("size", _.size).addHints(smithy.api.Default(smithy4s.Document.fromDouble(0.0d)), smithy.api.Required(), smithy.api.HttpHeader("X-Size")),
+    ObjectSize.schema.required[GetObjectOutput]("size", _.size).addHints(smithy.api.Default(smithy4s.Document.fromDouble(0.0d)), smithy.api.HttpHeader("X-Size"), smithy.api.Required()),
```
Comment (Baccata) [inline modules/example/src/smithy4s/example/GetObjectOutput.scala]: mmm, I wonder whether the CI is gonna scream about this : the ordering might be subject to the JVM you run. We'll see

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/ResourceSpec.scala (hunk 4 above)
Comment (Baccata) [reply modules/dynamic/test/src-jvm/smithy4s/dynamic/ResourceSpec.scala]: I'm happy enough with this tbh, the bit that was missing was the transitive gathering of operations, the fact that operations are correctly encoded dynamically is already tested.

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/OperationSpec.scala (hunk 2 above)
Comment (Baccata) [reply modules/dynamic/test/src-jvm/smithy4s/dynamic/OperationSpec.scala:35]: good catch

File: modules/example/src/smithy4s/example/GetObjectOutput.scala (hunk 8 above)
Comment (msosnicki) [reply modules/example/src/smithy4s/example/GetObjectOutput.scala]: Failed, when running `sbt test` from nix shell, it produces old output. Reverting

**Review by Baccata — COMMENTED** [review]
LGTM. @kubukoz, you wanna double check ?

**Review by Baccata — APPROVED** [review]

**Comment by kubukoz** [comment]
will do

File: build.sbt (hunk 5 above)
Comment (kubukoz) [reply build.sbt:207]: not saying to do it immediately, but any reason not to pull in the whole directory?

File: modules/dynamic/smithy/dynamic.smithy (hunk 6 above)
Comment (kubukoz) [reply modules/dynamic/smithy/dynamic.smithy:192]: guess we can follow up with that

**Review by kubukoz — APPROVED** [review]

**Pull request merged**